app/waf/src/controllers/adc.controller.ts

```diff
@@ -69,6 +61,7 @@ export class AdcController extends BaseController {
     private logger = factory.getLogger('controllers.adc'),
   ) {
     super(reqCxt);
+    this.tdMgr = new TrustedDeviceManager(this.trustedDeviceService);
```

Comment:
In fact, I do not know how to init this service object...

Reply:
Refer to `await new ComputeServiceProvider().value()`?

Reply:
Maybe it is not allowed to call await in constructor...

Reply:
Yeah, I did it with `TrustedDeviceManager.instanlize()` and putting `await new TustedDeviceServiceProvider().value()` inside instanlize.

Reply:
So you mean to call another init function in rest handler?

Reply:
I think you can refer other *ServiceProvider. The aim is encapsulating `TustedDeviceService` away. Since you have *Manager, it doesn't belong to this code context/level.

Reply:
If we attempt to encapsulate Service initialization steps, that means we have to do that by ourselves. It may 'look better' for somebody, but it may not be a better solution. Service object initialization is done by the framework, those are transparent to app developer. We can just consume the ready to use Service object. This TrustedDeviceManager is just to encapsulate some of the function code which related this service, so that Controller code may looks a little simpler. It may not be responsible to do the things in framework scope. If the solution make the code become complex, as you did for OpenStack integration, the cost for review and testing will increase. It may also increase the future development and maintain investment. I will select a way to make the program become simpler, so that other persons can easily understand and improve. I will not select a way to make the program complex.

Reply:
I think from the code design aspect, `TustedDeviceService` is used by `TrustedDeviceManager` to service `AdcController`, In fact, it has no relationship with the controller. From this point, removing it from the controller is good design. 
Programming simplification should be firstly the design simplification.  

I think You'd like to leave it in controller just because you don't know how to initialize it by yourself. But that doesn't stop we can have a clear design. 
If we think `await new TustedDeviceServiceProvider().value()` complex, correct thinking is to find why we cannot initialize it like the controller, instead of escaping of thinking, totally depending on the ts framework.

Reply:
I did not find a very good place to this 'await'. If we plan encapsulate  to this 'await' in one of member function of TrustedDeviceManager, we may need to call it in controller handler. or call it during another 'Promise' place like the 'openstack component'. I think that would not be a simple and cheap solution. A very easy way is to continue to consume the service object init by the controller.

Reply:
Not sure which `await` do you mean. 
`await TrustedDeviceManager.instanlize()`, do you mean this await?

Reply:
await new TustedDeviceServiceProvider().value()

Reply:
ok, let's not too focus on this point since I've known the code's state/logic. 
You can keep it as it's now.